From f3fe2f990bbafae88206c31885f5d4d8f168b19f Mon Sep 17 00:00:00 2001
From: Todd Tannenbaum <tannenba@cs.wisc.edu>
Date: Thu, 4 Mar 2021 07:55:53 -0600
Subject: [PATCH] (HTCONDOR-315) Always handle exceptions when decoding JWT
 tokens.

(modified by Matyas Selmeci <matyas@cs.wisc.edu> to fix rejects)
---
 src/condor_io/condor_auth_passwd.cpp  | 12 ++++++++----
 src/condor_utils/condor_scitokens.cpp | 11 +++++++++--
 2 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/src/condor_io/condor_auth_passwd.cpp b/src/condor_io/condor_auth_passwd.cpp
index e456ed2..6be0780 100644
--- a/src/condor_io/condor_auth_passwd.cpp
+++ b/src/condor_io/condor_auth_passwd.cpp
@@ -583,10 +583,14 @@ Condor_Auth_Passwd::fetchLogin()
 					}
 					else {
 						username = identity;
-						auto decoded_jwt = jwt::decode(local_token);
-						signature = decoded_jwt.get_signature();
-						token = decoded_jwt.get_header_base64() + "." + decoded_jwt.get_payload_base64();
-						found_token = true;
+						try {
+							auto decoded_jwt = jwt::decode(local_token);
+							signature = decoded_jwt.get_signature();
+							token = decoded_jwt.get_header_base64() + "." + decoded_jwt.get_payload_base64();
+							found_token = true;
+						} catch (...) {
+							dprintf(D_SECURITY, "Generated pool_password token is malformed\n");
+						}
 					}
 				} else {
 					dprintf(D_SECURITY, "No compatible security key found.\n");
diff --git a/src/condor_utils/condor_scitokens.cpp b/src/condor_utils/condor_scitokens.cpp
index c6a71ca..e95db38 100644
--- a/src/condor_utils/condor_scitokens.cpp
+++ b/src/condor_utils/condor_scitokens.cpp
@@ -90,8 +90,15 @@ htcondor::validate_scitoken(const std::string &scitoken_str, std::string &issuer
 	}
 
 	if (ident && IsDebugCategory(D_AUDIT)) {
-		auto jwt = jwt::decode(scitoken_str);
-		dprintf(D_AUDIT, ident, "Examining SciToken with payload %s.\n", jwt.get_payload().c_str());
+		try {
+			auto jwt = jwt::decode(scitoken_str);
+			dprintf(D_AUDIT, ident, "Examining SciToken with payload %s.\n", jwt.get_payload().c_str());
+		} catch (...) {
+			err.pushf("SCITOKENS", 2, "Failed to decode scitoken for audit log.");
+			dprintf(D_AUDIT, ident, "Failed to decode a SciToken in order to examine it - rejecting.\n");
+			return false;
+
+		}
 	}
 
 	SciToken token = nullptr;
-- 
2.6.3

